crates/vm/src/arch/segment_new.rs

```diff
@@ -0,0 +1,442 @@
+// TODO: Only use AddressMap/PagedVec
+//       Add metric counter for gas metering, figure out what else to add
+//       Proper benchmarking, profiling - close feedback loop
+use backtrace::Backtrace;
+use openvm_instructions::{
+    exe::FnBounds,
+    instruction::{DebugInfo, Instruction},
+    program::Program,
+};
+use openvm_stark_backend::{
+    config::{Domain, StarkGenericConfig},
+    keygen::types::LinearConstraint,
+    p3_commit::PolynomialSpace,
+    p3_field::PrimeField32,
+    prover::types::{CommittedTraceData, ProofInput},
+    utils::metrics_span,
+    Chip,
+};
+
+use super::{
+    ExecutionError, GenerationError, Streams, SystemConfig, VmChipComplex, VmComplexTraceHeights,
+    VmConfig,
+};
+#[cfg(feature = "bench-metrics")]
+use crate::metrics::VmMetrics;
+use crate::{
+    arch::{instructions::*, ExecutionState, InstructionExecutor},
+    system::memory::{online::GuestMemory, MemoryImage},
+};
+
+/// Check segment every 100 instructions.
+const SEGMENT_CHECK_INTERVAL: usize = 100;
+
+/// Prepared instruction type for efficient execution
+pub type PInstruction = Instruction<u32>;
+
+/// Represents the state of the VM during execution
+pub struct VmExecutionState<Mem, Ctx>
+where
+    Mem: GuestMemory,
+{
+    /// Current timestamp representing clock cycles
+    pub timestamp: u32,
```

Comment:
if this refers to clock cycles, let's name it something other than `timestamp` since timestamp refers to a specific thing in most of the system

Reply:
should this be in `Ctx`? not sure